File: examples/hello-world/gabriel-greeting-app-kotlin-compose/app/src/main/kotlin/org/organicprogramming/gabriel/greeting/kotlincompose/runtime/HolonCatalog.kt
package org.organicprogramming.gabriel.greeting.kotlincompose.runtime

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.organicprogramming.gabriel.greeting.kotlincompose.model.GabrielHolonIdentity
import org.organicprogramming.holons.BUILT
import org.organicprogramming.holons.CACHED
import org.organicprogramming.holons.CWD
import org.organicprogramming.holons.Discover
import org.organicprogramming.holons.INSTALLED
import org.organicprogramming.holons.LOCAL
import org.organicprogramming.holons.NO_LIMIT
import org.organicprogramming.holons.NO_TIMEOUT
import org.organicprogramming.holons.SIBLINGS

interface HolonCatalog {
    suspend fun discover(): List<GabrielHolonIdentity>
}

class DesktopHolonCatalog : HolonCatalog {
    override suspend fun discover(): List<GabrielHolonIdentity> = withContext(Dispatchers.IO) {
        AppPaths.configureRuntimeEnvironment()
        val result = Discover(
            LOCAL,
            null,
            AppPaths.discoveryRoot(),
            SIBLINGS or CWD or BUILT or INSTALLED or CACHED,
            NO_LIMIT,
            NO_TIMEOUT,
        )

        if (!result.error.isNullOrBlank()) {
            throw IllegalStateException(result.error)
        }

        buildMap<String, GabrielHolonIdentity> {
            result.found.forEach { ref ->
                val holon = GabrielHolonIdentity.fromDiscovered(ref) ?: return@forEach
                putIfAbsent(holon.slug, holon)
            }
        }.values.sortedWith(
            compareBy<GabrielHolonIdentity> { it.sortRank }
                .thenBy { it.displayName.lowercase() },
        )
    }
}

File: examples/hello-world/gabriel-greeting-app-flutter/app/lib/src/runtime/holon_catalog.dart
import 'package:collection/collection.dart';
import 'package:holons/holons.dart' as holons;

import '../model/app_model.dart';

abstract interface class HolonCatalog {
  Future<List<GabrielHolonIdentity>> discover();
}

class DesktopHolonCatalog implements HolonCatalog {
  @override
  Future<List<GabrielHolonIdentity>> discover() async {
    final result = holons.Discover(
      holons.LOCAL,
      null,
      null,
      holons.ALL,
      holons.NO_LIMIT,
      holons.NO_TIMEOUT,
    );

    if (result.error != null && result.error!.isNotEmpty) {
      throw StateError(result.error!);
    }

    final deduped = <String, GabrielHolonIdentity>{};
    for (final ref in result.found) {
      final holon = GabrielHolonIdentity.fromDiscovered(ref);
      if (holon == null) {
        continue;
      }
      deduped.putIfAbsent(holon.slug, () => holon);
    }

    return deduped.values.sorted((left, right) {
      if (left.sortRank != right.sortRank) {
        return left.sortRank.compareTo(right.sortRank);
      }
      return left.displayName.toLowerCase().compareTo(
        right.displayName.toLowerCase(),
      );
    });
  }
}
